### PR TITLE
List: set aria-setsize and aria-posinset

### LINKED ISSUE #27900: Tree: set aria-setsize and aria-posinset
The tree/list is virtualized and as such a screen reader cannot make a correct statement on the number of entries. We should leverage [aria-setsize](https://www.w3.org/TR/wai-aria/states_and_properties#aria-setsize) and [aria-posinset](https://www.w3.org/TR/wai-aria/states_and_properties#aria-posinset) for that purpose within the tree and list widgets. The widgets know the number of elements to render and the currently selected items position in the set.

/cc @joaomoreno @isidorn

### PR DESCRIPTION
**Pull request by isidorn**
Fixes #27900 for the list widget

Please note: that the `splice` seems to be called even when the focus changes in the list so I might be calling set `aria-setsize` a bit too often.

Also note that `aria-posinset` should start with 1, not 0 thus I am doing +1. More details: https://www.w3.org/TR/wai-aria/states_and_properties#aria-posinset

@joaomoreno once you approve on this I can do a similar thing for the tree
@alexandrudima fyi

### CODE REVIEW

File: src/vs/base/browser/ui/list/listWidget.ts [hunk 1]
```diff
@@ -656,7 +656,10 @@ export class List<T> implements ISpliceable<T>, IDisposable {
 	}
 
 	splice(start: number, deleteCount: number, elements: T[] = []): void {
-		this.eventBufferer.bufferEvents(() => this.spliceable.splice(start, deleteCount, elements));
+		this.eventBufferer.bufferEvents(() => {
+			this.spliceable.splice(start, deleteCount, elements);
+			this.view.domNode.setAttribute('aria-setsize', this.length.toString());
```
Comment (joaomoreno) [inline src/vs/base/browser/ui/list/listWidget.ts]: You should do this outside the event buffering.

Shouldn't there be an initial value for this, before any `splice` is called?

File: src/vs/base/browser/ui/list/listWidget.ts [hunk 2]
```diff
@@ -712,6 +715,7 @@ export class List<T> implements ISpliceable<T>, IDisposable {
 	setFocus(indexes: number[]): void {
 		indexes = indexes.sort(numericSort);
 		this.focus.set(indexes);
+		this.view.domNode.setAttribute('aria-posinset', indexes.length ? (indexes[0] + 1).toString() : undefined);
```
Comment (joaomoreno) [inline src/vs/base/browser/ui/list/listWidget.ts]: The [w3 docs](https://www.w3.org/TR/wai-aria/states_and_properties#aria-posinset) state:

> Authors MUST set the value for aria-posinset to an integer greater than or equal to 1, and less than or equal to the size of the set.

Setting it to `undefined` doesn't sound like it's the right thing to do here.

**Review by joaomoreno — CHANGES_REQUESTED** [review]

File: src/vs/base/browser/ui/list/listWidget.ts (hunk 1 above)
Comment (isidorn) [inline src/vs/base/browser/ui/list/listWidget.ts]: Initial value is 0 and I can set it in the constructor.
Ok so I will do it in the method ouside of the buffer

File: src/vs/base/browser/ui/list/listWidget.ts (hunk 2 above)
Comment (isidorn) [inline src/vs/base/browser/ui/list/listWidget.ts]: In that case I will jost not set it but remove the attribute, I believe that is the right thing to do

File: src/vs/base/browser/ui/list/listWidget.ts [hunk 3]
```diff
@@ -651,12 +651,16 @@ export class List<T> implements ISpliceable<T>, IDisposable {
 		if (options.ariaLabel) {
 			this.view.domNode.setAttribute('aria-label', options.ariaLabel);
 		}
+		this.view.domNode.setAttribute('aria-setsize', this.length.toString());
```
Comment (joaomoreno) [inline src/vs/base/browser/ui/list/listWidget.ts]: Just make this `0`, to be explicit.

File: src/vs/base/browser/ui/list/listWidget.ts [hunk 4]
```diff
@@ -651,12 +651,16 @@ export class List<T> implements ISpliceable<T>, IDisposable {
...
 
 		this.style(options);
 	}
 
 	splice(start: number, deleteCount: number, elements: T[] = []): void {
-		this.eventBufferer.bufferEvents(() => this.spliceable.splice(start, deleteCount, elements));
+		this.eventBufferer.bufferEvents(() => {
```
Comment (joaomoreno) [inline src/vs/base/browser/ui/list/listWidget.ts]: Return this to the old syntax.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Looks good, you can merge after last changes.

**Comment by alexandrudima** [comment]
Nice! Thank you @isidorn @joaomoreno !

**Pull request merged**